eyecite/models.py

```diff
@@ -41,75 +44,65 @@ def includes_year(
         )
 
 
-class Citation:
+@dataclass(eq=True, unsafe_hash=True)
+class CitationBase:
+    """Base class for objects returned by get_citations()."""
+
+    token: "Token"  # token this citation came from
+    index: int  # index of _token in the token list
+
+    def matched_text(self):
+        """Text that identified this citation, such as '1 U.S. 1' or 'Id.'"""
+        return str(self.token)
+
+    def span(self):
+        """Start and stop offsets in source text for matched_text.()"""
+        return (self.token.start, self.token.end)
```

Comment:
Can the `IdCitation` class override this method to somehow extend the calculated `end` value to include the end of its `after_tokens`, if its `has_page == True`? (Idea being that something like "Id., at 123" should be annotated fully, not just the "Id." part.)

Reply:
Same thing with `SupraCitation` actually, with its `antecedent_guess` and possible `page`?

Reply:
Huh, yeah. Id with has_page=True makes sense to me; antecedent_guess strikes me as too noisy. I can poke at it.

Reply:
I should also double check that annotate can handle overlapping annotations.

Reply:
Added handling for IdCitation, and a test showing `<0>1 U.S. 1</0>. Foo. <1>Id.</1> Bar. <2>Id. at 2.</2>`. 

I didn't take on SupraCitation, which currently looks both over-inclusive and under-inclusive to me in setting `page` and `antecedent_guess`. I'd suggest saving that for a separate pass that includes seeing if those need to be tweaked based on what they're currently matching.

One implementation detail -- we have to manually calculate the end offset in `IdCitation.span` by adding up the length of `after_tokens` strings, because regular strings aren't `Token` objects. It might be worth taking a performance hit to have everything be a token with offsets tracked rather than do that fussy math. I'd also like to save that for a future look though.

Reply:
Yeah, I agree that SupraCitation is problematic, probably best to leave that for latter.